### ISSUE TITLE
No error is reported on `EmbeddedAttribute` with `file` accessibility

### ISSUE THREAD

**Issue by ltrzesniewski** [dotnet/roslyn#80105]
Declaring `Microsoft.CodeAnalysis.EmbeddedAttribute` with a `file` accessibility:

- Does not produce error `CS9271`, which states the attribute must be `internal`
- The attribute is renamed in the compiled assembly
- It is therefore not taken into account in a consuming project, whose name is indicated by a `[InternalsVisibleTo]`

This is dangerous, as it *seems* to work at first glance since the project compiles, but the build can fail in referencing projects.

This [has been noticed](https://github.com/JetBrains/JetBrains.Annotations/issues/45#issuecomment-3243345946) by @controlflow, I'm just providing the repro. 🙂 

**Version Used**:

- Compiler version: '5.0.0-2.25380.108 (30000d88)'.
- SDK: 10.0.100-preview.7.25380.108

**Steps to Reproduce**:

Build the following solution: [EmbeddedAttribute.zip](https://github.com/user-attachments/files/22101091/EmbeddedAttribute.zip)

Here's the code from the zip (reformatted a bit for readability):

**`ProjectA`** - Defines and uses `EmbeddedAttribute` on an internal `StringSyntaxAttribute` visible to `ProjectB`

```xml
<Project Sdk="Microsoft.NET.Sdk">
  <PropertyGroup>
    <TargetFramework>netstandard2.0</TargetFramework>
    <LangVersion>13.0</LangVersion>
  </PropertyGroup>
</Project>
```

```C#
using System;
using System.Runtime.CompilerServices;
using Microsoft.CodeAnalysis;

[assembly: InternalsVisibleTo("ProjectB")]

namespace Microsoft.CodeAnalysis
{
    file sealed partial class EmbeddedAttribute : Attribute; //  <---- Issue here
}

namespace System.Diagnostics.CodeAnalysis
{
    [Embedded]
    [AttributeUsage(AttributeTargets.Parameter | AttributeTargets.Field | AttributeTargets.Property)]
    internal sealed class StringSyntaxAttribute : Attribute
    {
        public StringSyntaxAttribute(string syntax)
        {
        }
    }
}
```

**`ProjectB`** - References `ProjectA` and uses `StringSyntaxAttribute`

```xml
<Project Sdk="Microsoft.NET.Sdk">
  <PropertyGroup>
    <TargetFramework>net10.0</TargetFramework>
  </PropertyGroup>
  <ItemGroup>
    <ProjectReference Include="..\ProjectA\ProjectA.csproj" />
  </ItemGroup>
</Project>
```

```C#
using System.Diagnostics.CodeAnalysis;

namespace ProjectB;

public class Test
{
    public void Method([StringSyntax("Regex")] string value)
    {
    }
}
```

[Here's the repro of `ProjectA`](https://lab.razor.fyi/#hVHNShxBECZiJOlT8AmaOe1CmAdwT6tJIAFhcNYIipCantqxQk93U1UTMiT7BnmJnH2DgBffS5CddVcxC_ap6-erqu_7zJ_XxhQcG4Y2d7J_t9sJhcaWvSi2E_M0yj8QNCGKkpP8KNY4DeB7IXnWdtIFpRbzo9gm8sgl8g9yuGk7JsdR4lyfDTEXIIJt5fsD-zkocgAvX0mo8jiLo6zg-B2dHmbjS2MCtCgJHL50nPllrLX24mNbYV1jfbkKp6pMVad4KtDgaBPOgBtUyQtgaFGR7W_7X_EToa-3FQqOCVn78WoLPbCwguCxts6DiC2VKTRlHxR-bibYg8dpA3Z19vKlrvLktsNGMmStDOnxBvOIXgy_hVk81Wy7BWupyi4lRpFjlEGc7ATLK-CEnL23WQGsBH7WJzwjvSopNB6XyaUvS_ycPK4Zp1XzA_O1B1tZT8zC7O95UBT9suPk_O2b25vrf3_n73a_vboH) in .NET Lab, which shows how the attribute is renamed. Note that it is still applied onto itself.

**Diagnostic Id**:

- Expected: `CS9271` in `ProjectA`
- Actual: `CS0433` in `ProjectB`

**Expected Behavior**:

- `ProjectA` should not compile, with error `CS9271`:

  > Error CS9271 : The type 'Microsoft.CodeAnalysis.EmbeddedAttribute' must be non-generic, internal, sealed, non-static, have a parameterless constructor, inherit from System.Attribute, and be able to be applied to any type.

**Actual Behavior**:

- `ProjectA` compiles just fine

- `EmbeddedAttribute` is not taken into account in `ProjectB` (probably because the attribute has been renamed), which generates error `CS0433`:

  > Error CS0433 : The type 'StringSyntaxAttribute' exists in both 'ProjectA, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null' and 'System.Runtime, Version=10.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a'

- The compiled `ProjectA` assembly contains:

  ```C#
  namespace Microsoft.CodeAnalysis;

  [<CodeAnalysis>F227F589E3E1DDA51C2ACFB7A4006583F22A638EF7F02E5263C80EC3710A5DC3B__Embedded]
  internal sealed class <CodeAnalysis>F227F589E3E1DDA51C2ACFB7A4006583F22A638EF7F02E5263C80EC3710A5DC3B__EmbeddedAttribute : Attribute;
  ```

- I noticed that in this example, `ProjectA` also defines a "real" `EmbeddedAttribute` marked as `[CompilerGenerated]`, which is used on a generated `System.Runtime.CompilerServices.RefSafetyRulesAttribute`.

**Comment by RikkiGibson** [comment]
Strictly a file-local declaration really does have internal accessibility. The feature is spec'd and implemented as a concept independent from accessibility. But it does change the metadata name of the type as you noted.

Does this seem like something we should have blocked in #76523? @333fred.

We might want set a precedent, that in the case of well-known attributes which have downstream effects on the usages of the attributed members, that the attribute type declarations can't be `file`-local.

**Comment by 333fred** [comment]
> Does this seem like something we should have blocked in https://github.com/dotnet/roslyn/pull/76523? @333fred.

Yes, we should fix this.

**Comment by ltrzesniewski** [comment]
> Strictly a file-local declaration really does have internal accessibility. The feature is spec'd and implemented as a concept independent from accessibility.

That's an implementation detail: `file` is [documented](https://learn.microsoft.com/en-us/dotnet/csharp/programming-guide/classes-and-structs/access-modifiers) as an accessibility level, just like the other ones. The user is not supposed to know how this is implemented.



> We might want set a precedent, that in the case of well-known attributes which have downstream effects on the usages of the attributed members, that the attribute type declarations can't be `file`-local.

I can see two solutions for `EmbeddedAttribute`, at least:

- Disallow `file`, like the `CS9271` message currently says
- Allow `file`, but turn it into `internal sealed partial`

The first solution is straightforward and is basically what this issue asks for.

The second solution could be implemented later as a new feature without breaking changes if you decide it would be helpful to the users. `EmbeddedAttribute` is already treated very specially by the compiler, so allowing `file` might even fit.

**Comment by RikkiGibson** [comment]
https://github.com/dotnet/csharplang/blob/main/proposals/csharp-11.0/file-local-types.md#accessibility

> The `file` modifier is not classified as an accessibility modifier. No accessibility modifiers can be used in combination with `file` on a type. `file` is treated as an independent concept from accessibility. Since file-local types can't be nested, only the default accessibility `internal` is usable with `file` types.

It looks like a miscommunication happened with docs team. Opened dotnet/docs#48224.
